lib/ansible/modules/files/unarchive.py

```diff
@@ -796,7 +796,8 @@ def main():
         # If remote_src=true, and src= contains ://, try and download the file to a temp directory.
         elif '://' in src:
             tempdir = os.path.dirname(os.path.realpath(__file__))
-            package = os.path.join(tempdir, str(src.rsplit('/', 1)[1]))
+            tempfile = re.search(r'\/([^/?]*)(\?.*)?$', src)[1]
+            package = os.path.join(tempdir, str(tempfile))
```

Comment:
Instead of using the url to try and determine the name of a tempfile, let's use real tempfile functionality instead.

The `get_url` module already uses this technique, so we should do it here too:

https://github.com/ansible/ansible/blob/05c6ff79f9860dbd6b43cb4914ee749baf65b9f7/lib/ansible/modules/net_tools/basics/get_url.py#L375